File: app/src/main/java/com/stevenfrew/beatprompter/BeatPrompterApplication.java
package com.stevenfrew.beatprompter;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.AssetManager;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.support.multidex.MultiDex;

import com.stevenfrew.beatprompter.bluetooth.BluetoothManager;
import com.stevenfrew.beatprompter.event.CancelEvent;
import com.stevenfrew.beatprompter.midi.MIDIController;
import com.stevenfrew.beatprompter.songload.SongLoadInfo;
import com.stevenfrew.beatprompter.songload.SongLoaderTask;

public class BeatPrompterApplication extends Application {
    public static final String TAG = "beatprompter";
    public final static String APP_NAME="BeatPrompter";
    private static Application mApp;
    private final static String SHARED_PREFERENCES_ID="beatPrompterSharedPreferences";
    private static SongLoaderTask mSongLoaderTask = new SongLoaderTask();
    private Thread mSongLoaderTaskThread = new Thread(mSongLoaderTask);

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    public void onCreate() {
        super.onCreate();
        mApp=this;
        MIDIController.initialise(this);
        BluetoothManager.initialise(this);
        mSongLoaderTaskThread.start();
        Task.resumeTask(mSongLoaderTask);
    }

    public void onTerminate()
    {
        Task.stopTask(mSongLoaderTask,mSongLoaderTaskThread);
        BluetoothManager.shutdown(this);
        MIDIController.shutdown(this);
        super.onTerminate();
    }

    public static String getResourceString(int resID)
    {
        return mApp.getString(resID);
    }

    public static String getResourceString(int resID,Object... args)
    {
        return mApp.getString(resID,args);
    }

    public static SharedPreferences getPreferences()
    {
        return PreferenceManager.getDefaultSharedPreferences(mApp);
    }

    public static SharedPreferences getPrivatePreferences()
    {
        return mApp.getSharedPreferences(SHARED_PREFERENCES_ID, Context.MODE_PRIVATE);
    }

    public static AssetManager getAssetManager()
    {
        return mApp.getAssets();
    }

    public static Context getContext()
    {
        return mApp.getApplicationContext();
    }

    public static void loadSong(SongLoadInfo sli, Handler handler, CancelEvent cancelEvent, boolean registered)
    {
        mSongLoaderTask.loadSong(sli,handler,cancelEvent,registered);
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/BeatPrompterApplication.kt
package com.stevenfrew.beatprompter

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.content.res.AssetManager
import android.preference.PreferenceManager
import android.support.multidex.MultiDex
import com.stevenfrew.beatprompter.comm.bluetooth.BluetoothManager
import com.stevenfrew.beatprompter.comm.midi.MIDIController
import com.stevenfrew.beatprompter.song.load.SongLoadQueueWatcherTask

class BeatPrompterApplication : Application() {
    private val mSongLoaderTaskThread = Thread(SongLoadQueueWatcherTask)

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    override fun onCreate() {
        super.onCreate()
        mApp = this
        MIDIController.initialise(this)
        BluetoothManager.initialise(this)
        mSongLoaderTaskThread.start()
        Task.resumeTask(SongLoadQueueWatcherTask)
    }

    override fun onTerminate() {
        Task.stopTask(SongLoadQueueWatcherTask, mSongLoaderTaskThread)
        BluetoothManager.shutdown(this)
        MIDIController.shutdown(this)
        super.onTerminate()
    }

    companion object {
        const val TAG = "beatprompter"
        const val TAG_LOAD = "beatprompter_load"
        const val TAG_COMMS = "beatprompter_comms"
        const val APP_NAME = "BeatPrompter"
        private lateinit var mApp: Application
        private const val SHARED_PREFERENCES_ID = "beatPrompterSharedPreferences"

        fun getResourceString(resID: Int): String {
            return mApp.getString(resID)
        }

        fun getResourceString(resID: Int, vararg args: Any): String {
            return mApp.getString(resID, *args)
        }

        val preferences: SharedPreferences
            get() = PreferenceManager.getDefaultSharedPreferences(mApp)

        val privatePreferences: SharedPreferences
            get() = mApp.getSharedPreferences(SHARED_PREFERENCES_ID, Context.MODE_PRIVATE)

        val assetManager: AssetManager
            get() = mApp.assets

        val context: Context
            get() = mApp.applicationContext
    }
}
